Service with Connect sidecar de-registered without removing sidecar first

Nomad implicitly registers Connect sidecar services via the `connect.sidecar_service` stanza, and does not keep track of the sidecar service or its checks internally. When it's time to shut the task down, Nomad only de-registers the parent service (because that's all it knows about). This causes an ERROR and WARN in Consul logs - it's the root cause of https://github.com/hashicorp/consul/issues/8657

Nomad needs to explicitly de-register the sidecar service before de-registering the parent service to avoid making Consul unhappy. 

```
2021-07-02T10:14:55.473-0500 [WARN]  agent.cache: handling error in Cache.Notify: cache-type=service-http-checks error="Internal cache failure: service '_nomad-task-4903435d-3694-9ebb-21ca-378815a1d366-group-api-count-api-9001' not in agent state" index=0
2021-07-02T10:14:55.473-0500 [ERROR] agent.proxycfg: Failed to handle update from watch: service_id=_nomad-task-4903435d-3694-9ebb-21ca-378815a1d366-group-api-count-api-9001-sidecar-proxy id=service-http-checks:_nomad-task-4903435d-3694-9ebb-21ca-378815a1d366-group-api-count-api-9001 error="error filling agent cache: Internal cache failure: service '_nomad-task-4903435d-3694-9ebb-21ca-378815a1d366-group-api-count-api-9001' not in agent state"
2021-07-02T10:14:55.474-0500 [INFO]  agent: Deregistered service: service=_nomad-task-4903435d-3694-9ebb-21ca-378815a1d366-group-api-count-api-9001
2021-07-02T10:14:55.474-0500 [WARN]  agent.cache: handling error in Cache.Notify: cache-type=service-http-checks error="Internal cache failure: service '_nomad-task-4903435d-3694-9ebb-21ca-378815a1d366-group-api-count-api-9001' not in agent state" index=0
2021-07-02T10:14:55.474-0500 [ERROR] agent.proxycfg: Failed to handle update from watch: service_id=_nomad-task-4903435d-3694-9ebb-21ca-378815a1d366-group-api-count-api-9001-sidecar-proxy id=service-http-checks:_nomad-task-4903435d-3694-9ebb-21ca-378815a1d366-group-api-count-api-9001 error="error filling agent cache: Internal cache failure: service '_nomad-task-4903435d-3694-9ebb-21ca-378815a1d366-group-api-count-api-9001' not in agent state"
2021-07-02T10:14:55.474-0500 [INFO]  agent: Deregistered service: service=_nomad-task-4903435d-3694-9ebb-21ca-378815a1d366-group-api-count-api-9001-sidecar-proxy
```

--- a/.changelog/10873.txt
+++ b/.changelog/10873.txt
@@ -1,0 +1,3 @@
+```release-note:bug
+consul/connect: Fixed a bug where service deregistered before connect sidecar
+```

--- a/command/agent/consul/service_client.go
+++ b/command/agent/consul/service_client.go
@@ -821,8 +821,23 @@ func (c *ServiceClient) sync(reason syncReason) error {
 			continue
 		}
 
-		// Unknown Nomad managed service; kill
+		// Get the Consul namespace this service is in.
 		ns := servicesInConsul[id].Namespace
+
+		// If this service has a sidecar, we need to remove the sidecar first,
+		// otherwise Consul will produce a warning and an error when removing
+		// the parent service.
+		//
+		// The sidecar is not tracked on the Nomad side; it was registered
+		// implicitly through the parent service.
+		if sidecar := getNomadSidecar(id, servicesInConsul); sidecar != nil {
+			if err := c.agentAPI.ServiceDeregisterOpts(sidecar.ID, &api.QueryOptions{Namespace: ns}); err != nil {
+				metrics.IncrCounter([]string{"client", "consul", "sync_failure"}, 1)
+				return err
+			}
+		}
+
+		// Remove the unwanted service.
 		if err := c.agentAPI.ServiceDeregisterOpts(id, &api.QueryOptions{Namespace: ns}); err != nil {
 			if isOldNomadService(id) {
 				// Don't hard-fail on old entries. See #3620

